src/ir/LocalStructuralDominance.cpp

```diff
@@ -0,0 +1,231 @@
+            }
+            return;
+          }
+          case Expression::Id::LocalSetId: {
+            auto* set = curr->cast<LocalSet>();
+            if (!self->localsSet[set->index]) {
+              self->pushTask(doLocalSet, currp);
+            }
+            // Immediately continue in the loop.
+            currp = &set->value;
+            continue;
+          }
+
+          // Control flow structures.
+          case Expression::Id::BlockId: {
+            auto* block = curr->cast<Block>();
+            // Blocks with no name are never emitted in the binary format, so do
+            // not create a scope for them.
+            if (block->name.is()) {
+              self->pushTask(Scanner::doEndScope, currp);
+            }
+            auto& list = block->list;
+            for (int i = int(list.size()) - 1; i >= 0; i--) {
+              self->pushTask(Scanner::scan, &list[i]);
+            }
+            if (block->name.is()) {
+              // Just call the task immediately.
+              doBeginScope(self, currp);
+            }
+            return;
+          }
+          case Expression::Id::IfId: {
+            if (curr->cast<If>()->ifFalse) {
+              self->pushTask(Scanner::doEndScope, currp);
+              self->maybePushTask(Scanner::scan, &curr->cast<If>()->ifFalse);
+              self->pushTask(Scanner::doBeginScope, currp);
+            }
+            self->pushTask(Scanner::doEndScope, currp);
+            self->pushTask(Scanner::scan, &curr->cast<If>()->ifTrue);
+            self->pushTask(Scanner::doBeginScope, currp);
+            // Immediately continue in the loop.
+            currp = &curr->cast<If>()->condition;
+            continue;
+          }
+          case Expression::Id::LoopId: {
+            self->pushTask(Scanner::doEndScope, currp);
+            // Just call the task immediately.
+            doBeginScope(self, currp);
+            // Immediately continue in the loop.
+            currp = &curr->cast<Loop>()->body;
+            continue;
+          }
+          case Expression::Id::TryId: {
+            auto& list = curr->cast<Try>()->catchBodies;
+            for (int i = int(list.size()) - 1; i >= 0; i--) {
+              self->pushTask(Scanner::doEndScope, currp);
+              self->pushTask(Scanner::scan, &list[i]);
+              self->pushTask(Scanner::doBeginScope, currp);
+            }
+            self->pushTask(Scanner::doEndScope, currp);
+            // Just call the task immediately.
+            doBeginScope(self, currp);
+            // Immediately continue in the loop.
+            currp = &curr->cast<Try>()->body;
+            continue;
+          }
+
+          default: {
+            // Control flow structures have been handled. This is an expression,
+            // which we scan normally.
+            assert(!Properties::isControlFlowStructure(curr));
+            PostWalker<Scanner>::scan(self, currp);
+            return;
```

Comment:
Not sure what you mean? `Unary` is not needed in this processing. Or do you mean `LocalSet` which has a single child? - that is already handled in the loop here.

Reply:
Specifically, we could add this:

```
case Expression::Id::UnaryId: {
  // Immediately continue in the loop.
  currp = curr->cast<Unary>()->value;
  continue;
}
```

We could do this with any expression type that only has a single operand.

Reply:
I see, yes, that could be faster. I think it would add a bunch of extra code though, and need to be updated when we add new instructions, etc. I am also unsure how much it would help. I did measure something related, btw - I checked whether we pushed a single child on the stack (by measuring the size before and after) and if so, continued in the loop on that. That does what you said, but it does write and read from the stack, so it just saves the call. I saw no significant benefit there.

Reply:
If the loop doesn't help much, would be simpler to do without it entirely?

Reply:
The loop does help in itself, in my measurements. Just adding those particular additional uses of the loop didn't help.

One reason the loop helps in the current code is that not only do we avoid a call, but also we can avoid pushing a task and later popping it and doing an indirect call. (see e.g. how Loop just calls `doBeginScope` rather than push it)

Reply:
I'm confused about why avoiding that task push, pop, and indirect call doesn't help as much when applied to other expression types. The existing code lgtm, though.

Reply:
Yeah, it's not totally obvious to me either. It could depend on the frequency of the instructions or their size (which affects caching) for example. Overall, the current code has some added optimization complexity but not much, and each piece seemed worthwhile when I measured it.